rfcs/proposed/task_group_dynamic_dependencies/README.md

```diff
@@ -0,0 +1,365 @@
+
+        for (tbb::task_handle& p : users::find_predecessors(work_id)) {
+            new_task.add_predecessor(p);
+        }
+
+        tg.run(new_task);
+    }
+
+Again, the graph, as shown below, is simple. However, now we do not 
+know the completion status of the predecessors. Therefore, 
+for ease-of-use, a `task_handle` should be usable as a dependency 
+regardless of state of the task it represents. Any predecessor 
+that is already completed when it is added as a predecessor will 
+not delay the start of the dependent task. Otherwise, end-users 
+will need to track these states explicitly.
+
+<img src="unknown_states.png" width=400>
+
+#### And example of recursive decomposition
+
+This example is a version of merge-sort (with many of the details left out). 
+Assume that there is an initial task that executes the function shown 
+below as its body, and the function implements that task, and also serves 
+as the body for the recursively decomposed pieces. The beginning and
+end of the sequence are represented by `b` and `e`, and much of the
+(unimportant) details of the implementation of merge-sort is hidden 
+behind the functions `users::do_serial_sort`, `users::create_left_range`,
+`users::create_right_range`, and `users::do_merge`.  
+
+    template<typename T>
+    void merge_sort(tbb::task_group& tg, T b, T e) {
+        if (users::range_is_too_small(b, e)) {
+            // base-case when range is small
+            users::do_serial_sort(b, e);
+        } else {
+            // calculate left and right ranges
+            T lb, le, rb, re;
+            users::create_left_range(lb, le, b, e);
+            users::create_right_range(rb, re, b, e);
+
+            // create the three tasks
+            tbb::task_handle sortleft =
+                tg.defer([lb, le, &tg] {
+                    merge_sort(tg, lb, le);
+                });
+            tbb::task_handle sortright =
+                tg.defer([rb, re, &tg] {
+                    merge_sort(tg, rb, re);
+                });
+            tbb::task_handle merge =
+            tg.defer([rb, re, &tg] {
+                    users::do_merge(tg, lb, le, rb, re);
+                });
+
+            // add predecessors for new merge task
+            merge.add_predecessors(sortleft, sortright);
+
+            // insert new subgraph between currently executing
+            // task and its successors
+            tbb::transfer_successors_to(merge);
+
+            tg.run(sortleft);
+            tg.run(sortright);
+            tg.run(merge);
+        }
+    }
+
+The task tree for this example matches the one shown earlier for merge-sort. 
+
+## Open Questions in Design
+
+Some open questions that remain:
+
+- Are the suggested APIs sufficient?
+- Are there additional use cases that should be considered that we missed in our analysis?
+- Are there other parts of the pre-oneTBB tasking API that developers have struggled to find a good alternative for?
```

Comment:
These are interesting questions.

While reading this RFC, I kind of rushed proposing additional syntax sugar that besides being more user-friendly, since it seems to represent popular use cases, can save some CPU cycles. So I am posting them here for a discussion.

1. Would it be useful to have a method that simultaneously joins (or even fuses?) the instantiation of a task and transferring of the current task successors?
Something like:
```cpp
    template <typename Func>
    task_handle transfer_successors_to(Func&& f);
```
For the recursive decomposition scenario, instantiating a new task within an executing task and transferring successors to that new task seems to be the main model of writing such algorithms. Although, it seems to be not saving much (only one call to the library and assign to a couple of pointers?), there is always(?) going to be such a sequence in the code. Otherwise, how else an execution of already submitted tasks can be postponed?

Shall we also consider a question of having that API _instead_ or _in addition_ to the one proposed?

2. As for `add_predecessor(s)` and `add_successor(s)` I have a couple of thoughts.
   **a.** It seems again that it might be useful to merge instantiation of the new task handles and adding them as successors/predecessors:
    ```cpp
        template <typename Func>
        void add_predecessor(Func&& f);
    
        template <typename Func>
        void add_successor(Func&& f);
    ```
   **b.** Also, I think having an API that would allow adding more than one predecessor/successor at once can be useful, since usually a number of successors/predecessors are instantiated. I only think that we don't need to limit ourselves to only two parameters as it was proposed optionally, but allow passing of an arbitrary size of task handles or even user lambdas. Of course, a pattern of having a single task producer might be viewed as a limiting one, but there actually might be the cases where tasks cannot be made available to the scheduler (i.e. spawned) until all of them are gathered together from different producers, which essentially represents a barrier in the execution pipeline. Not to mention that the spawning of a bunch of tasks all together are done faster than regular one by one spawnings. Spawning of a bunch of tasks at once was implemented in the old TBB, as far as I remember.
So here I suggest to have something like:
    ```cpp
        template <typename... Func>
        void add_predecessors(Func&& ...f);
    
        template <typename... Func>
        void add_successor(Func&& ...f);
    ```
    However, this also seems to ask for having the `task_group::run()` method to accept an arbitrary size of task handles and/or functors. So, perhaps, it is more related to another RFC/extension...

Reply:
I think we should start with minimal API then extend with syntactic sugar based on use cases. I suspect this will start as an experimental feature to allow some feedback on API. 

Even so, I'm also open to including these additional APIs in the initial implementation, if others think they're likely needed.